src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,318 @@
+{
+    "\\/:",
+    stdio_fopen,
+    stdio_fread,
+    stdio_fseek,
+    stdio_fclose,
+    stdio_ftell
+};
+
+struct UsfLoaderState
+{
+    bool enableCompare; // The _enablecompare tag is present in the file; passed to usf_set_compare
+    bool enableFIFOFull; // The _enableFIFOfull tag is present in the file; passed to usf_set_fifo_full
+    void * emu; // The emulator state
+};
+
+static int
+usf_loader(void * context, const uint8_t * exe, size_t exe_size,
+               const uint8_t * reserved, size_t reserved_size)
+{
+    struct UsfLoaderState * state = ( struct UsfLoaderState * ) context;
+
+    if ( exe && exe_size > 0 ) return -1;
+
+    return usf_upload_section( state->emu, reserved, reserved_size );
+}
+
+static int
+usf_info(void * context, const char * name, const char * value)
+{
+    struct UsfLoaderState * state = ( struct UsfLoaderState * ) context;
+    if (!value) return 0;
+
+    if ( strcmp( name, "_enablecompare" ) == 0)
+        state->enableCompare = 1;
+    else if ( strcmp( name, "_enableFIFOfull" ) == 0)
+        state->enableFIFOFull = 1;
+
+    return 0;
+}
+
+struct UsfTags {
+    double length = -1; // Track duration. -1 represents looping infinitely
+    double fade = 0;   // Fade out duration
+    bool setTags = false;  // If set, set the tags with the below variables
+    const TagHandler * tagHandler;
+    void * handler_ctx;
+};
+
+static double
+get_length_from_string(const char * string)
+{
+    size_t len = strlen(string);
+    int32_t total = 0; // Total time in milliseconds
+    int32_t finalMult = 1000; // Multiplier for the final unit. If no delimiter is used, use seconds.
+    int32_t localMult = 1; // Multiplier in the currently parsed unit (millisecond, second, minute).
+    int32_t tmp = 0; // 
+    for (int i = len - 1; i >= 0; i--) {
+        char c = string[i];
+        if (c >= '0' && c <= '9') {
+            tmp += (c - '0') * localMult;
+            localMult *= 10;
+        } else {
+            localMult = 1;
+            int mult = 0;
+            if (c == '.') {
+                mult = 1;
+            } else if (c == ':') {
+                mult = 1000;
+                finalMult = 60000;
+            } else {
+                return -1; // Error in parsing duration. Return -1 (looping).
+            }
+            total += tmp * mult;
+            tmp = 0;
+        }
+    }
+    total += finalMult*tmp;
+
+    return total/1000;
+}
+
+static bool
+usf_set_tag(const char * field, const char * name, const char *value, TagType type, UsfTags * tags)
+{
+    if (strcmp(name, field) == 0) {
+        tag_handler_invoke_tag(*tags->tagHandler, tags->handler_ctx, type, value);
+        return true;
+    }
+    return false;
+}
+
+static int
+usf_tags_target(void * context, const char * name, const char * value)
+{
+    struct UsfTags * tags = ( struct UsfTags * ) context;
+    if (!value) return 0;
+
+    if (tags->setTags) {
+        usf_set_tag("title", name, value, TAG_TITLE, tags) ||
```

Comment:
Are you suggesting I use std::map for this? Or would you rather have an if-else series mapping a string to a TagType?